induction/restriction, tensor product of representations for finite groups

jacksonwalters opened the pull request:
Include a class for constructing induced representations of the symmetric group, i.e. permutation modules $Ind_{S_k}^{S_n} V_{\lambda_k}$ where $V_{\lambda_k}$ is a Specht module corresponding to the partition $\lambda_k$.

Include a class for constructing restricted representations of the symmetric group of the form $Res_{S_k}^{S_n} V_{\lambda_n}$ where given a representation $V_{\lambda_n}$ of $S_n$ we get a representation of $S_k$ via $\rho(g)|_{S_k} = \rho(g)$ for $g \in S_k$ where $S_k \subset S_n$.

These representations should satisfy Frobenius reciprocity, that is $\langle Ind_{H}^{G} \psi, \varphi \rangle_{G} = \langle \psi, Res_{H}^{G} \varphi \rangle_{H}$ for any two class functions $\psi: H \rightarrow \mathbb{C}$ and $\varphi: G \rightarrow \mathbb{C}$.

We also include the external tensor product of representations $\rho_1 \boxtimes \rho_2$ of $S_{k_1} \times S_{k_2}$.

The goal is to compute induced representations of the Young subgroups $S_{\lambda_1} \times \ldots \times S_{\lambda_r} \subset S_n$ for a partition $\lambda$ of $n$. They should decompose in accordance with the Littlewood–Richardson rule, and restricted representations from $S_n$ to $S_{n-1}$ via the branching rules.

n.b. The Littlewood-Richardson coefficients $c_{\mu\nu}^{\lambda}$ can be computed quickly using Anders Buch's [`lrcalc`](https://sites.math.rutgers.edu/~asbuch/lrcalc/) package.

<!-- ^ Please provide a concise and informative title. -->
<!-- ^ Don't put issue numbers in the title, do this in the PR description below. -->
<!-- ^ For example, instead of "Fixes #12345" use "Introduce new method to calculate 1 + 2". -->
<!-- v Describe your changes below in detail. -->
<!-- v Why is this change required? What problem does it solve? -->
<!-- v If this PR resolves an open issue, please link to it here. For example, "Fixes #12345". -->



### :memo: Checklist

<!-- Put an `x` in all the boxes that apply. -->

- [x] The title is concise and informative.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [x] I have created tests covering the changes.
- [x] I have updated the documentation and checked the documentation preview.

### :hourglass: Dependencies

<!-- List all open PRs that this PR logically depends on. For example, -->
<!-- - #12345: short description why this is a dependency -->
<!-- - #34567: ... -->
[https://github.com/sagemath/sage/issues/41338](https://github.com/sagemath/sage/issues/41338)
[https://github.com/sagemath/sage/issues/41344](https://github.com/sagemath/sage/issues/41344)

jacksonwalters committed:
add induced/restricted/tensor product of rep'ns

add external tensor product of two representations of the symmetric group as a class.

remove whitespace

fix tensor product

issues: need to use an appropriate parent so that `self._n` can defined. it would also be good to include a method for computing dimensions using hook-length formula for representations.

use block_matrix

use the block_matrix constructor

add blank line

add induced/restricted representations

add induced representations

include a class for constructing induced representations of the symmetric group. in particular, permutation modules M^{\lambda} which are of the form Ind_{S_k}^{S_n} V_{\lambda_k} where V_{\lambda_k} is a Specht module corresponding to the partition \lambda_k.

fix bare exception

remove whitespace for linter

remove whitespace

remove carriage return

add string representation, subclass

add restricted representation

the restricted representation should be \rho(g) = \rho|_H(g) for a subgroup H < G. essentially, we act the subgroup of elements in H and forget about the action on G \ H. this yields a representation of H which may be reducible.

remove whitespace

verify Frobenius reciprocity, induction dimension

include code to provide basic verifications of Frobenius reciprocity and induction dimension.

remove dimension verification function

just perform the tests explicitly in the `TESTS::` section to avoid writing another function with its own doctests, etc.

use f-strings

remove whitespace

just use subgroups S_k of S_n

remove block_matrix

unnecessary

simplify, use Young subgroup S_k x S_{n-k}

avoid using try/except clauses. simplify and remove standalone functions.

remove whitespace

define `character_inner_product` as helper function

fix whitespace

make Frobenius reciprocity verification private

add missing import

clean up restrict

fix ExternalTensorProductRepresentation

fix InducedRepresentation

update

extend tensor product to multiple factors

slim down RestrictedRepresentation class

define dimension in InducedRepresentation

fix missing `_on_basis` definition

defines a default _on_basis function that acts on a basis element by embedding the subgroup element into the ambient group and delegating to the super-representation's action

begin revising RestrictedRepresentation

begin fixing Frobenius reciprocity

begin fixing TensorProductRepresentation

fix issues with RestrictedRepresentation

this is attempt to clean up the AI generated code for RestrictedRepresentation

constructed vector space for InducedRepresentation

moved InducedRepresentation code from symmetric_group_algebra

add imports

we need to move these to be constructed via the constructor factory function. this will require including additional optional parameters since these representations depend on more than just the partition, but rather a pair of representations, or representations to be induced, restricted along with the parameter for the subgroup, in this case just `k`.

import classes and functions explicitly

add TestSuite

add doctests

jacksonwalters committed:
add induced/restricted/tensor product of rep'ns

add external tensor product of two representations of the symmetric group as a class.

remove whitespace

fix tensor product

issues: need to use an appropriate parent so that `self._n` can defined. it would also be good to include a method for computing dimensions using hook-length formula for representations.

use block_matrix

use the block_matrix constructor

add blank line

add induced/restricted representations

add induced representations

include a class for constructing induced representations of the symmetric group. in particular, permutation modules M^{\lambda} which are of the form Ind_{S_k}^{S_n} V_{\lambda_k} where V_{\lambda_k} is a Specht module corresponding to the partition \lambda_k.

fix bare exception

remove whitespace for linter

remove whitespace

remove carriage return

add string representation, subclass

add restricted representation

the restricted representation should be \rho(g) = \rho|_H(g) for a subgroup H < G. essentially, we act the subgroup of elements in H and forget about the action on G \ H. this yields a representation of H which may be reducible.

remove whitespace

verify Frobenius reciprocity, induction dimension

include code to provide basic verifications of Frobenius reciprocity and induction dimension.

remove dimension verification function

just perform the tests explicitly in the `TESTS::` section to avoid writing another function with its own doctests, etc.

use f-strings

remove whitespace

just use subgroups S_k of S_n

remove block_matrix

unnecessary

simplify, use Young subgroup S_k x S_{n-k}

avoid using try/except clauses. simplify and remove standalone functions.

remove whitespace

define `character_inner_product` as helper function

fix whitespace

jacksonwalters committed:
import classes and functions explicitly

jacksonwalters commented:
We should use the constructor function rather than require importing the class every time it's used. This should be possible.

We need to modify the constructor function take some optional parameters. 

- For `InducedRepresentation`, we need `k` and the `sub_rep` of `S_k` we are inducing.
- For `RestrictedRepresentation`, we need `k` and the `representation` of `S_n` we are restricting.
- For `ExternalTensorProductRepresentation`, we need representations `rho1` and `rho2`. Then `n = `rho1._n+rho2._n`, so we don't need to pass it explicitly.

jacksonwalters committed:
add imports

we need to move these to be constructed via the constructor factory function. this will require including additional optional parameters since these representations depend on more than just the partition, but rather a pair of representations, or representations to be induced, restricted along with the parameter for the subgroup, in this case just `k`.

jacksonwalters commented:
@tscrim I'll refactor to use the constructor and get all tests passing ASAP, this week likely.

tscrim reviewed (changes requested):
Some quick comments:

- It would be better to implement the external tensor product in a more general setting and with an arbitrary number of factors. This can even use the generic `Representation` class as a base.
- Likewise, the restriction should be implemented very generically as well. Basically it just needs to know that its group is changed. The action comes for free.
- Ideally, we should just implement the induction in the general (finite group) case as well. However, this is somewhat expensive as we need to choose a set of coset representatives. If the group is implemented in GAP, then everything you need [seems to be there](https://docs.gap-system.org/doc/ref/chap39.html#X8412ABD57986B9FC). Then you need to figure out how to write an element as $g = rh \in rH \subseteq G$, where $r$ is the representative and $H \subseteq G$ is the subgroup inducing from.
- That said, for the case of Coxeter groups (which includes $S_n$) and parabolic subgroups, for each coset there is a special representative that has the minimal (Coxeter) length (for $S_n$, this is with the minimal set of inversions). So there is an algorithm to find such an elements $r$ and $h$ for any $g \in G$ by decreasing inversions (by right multiplication).
- The case of $S_k \subseteq S_n$, we are better thinking of this as the parabolic subgroup by $S_k \times S_1^{n-k}$. This is what you’re doing basically, but I am making this explicit because of the next point.
- For inducing $S_n$ representations that you’re after, it is better to give the input as a single partition *tuple* with no empty components of (total) size $n$. As an example/special case, a tuple of single rows is the so-called `TabloidRepresentation`. This is a very good special case to implement as a subclass.
- It’s best to not implement the matrices directly and let the base classes handle that for uniformity and simplicity (with less maintenance).
- The dimension of the induced representations is easy: $\dim I_H^G V = |G/H| \dim V = |G| / |H| \dim V$. You mentioned this but didn’t implement it (and the index of $S_k$ in $S_n$ is $n! / k! \neq \binom{n}{k}$!).
- You have a lot of details/formatting fix because AI doesn’t do a good job of writing code that matches SageMath’s (and sometimes Python’s) general conventions.

Everything except the induced representations should be very simple to take care of I think because of the current framework. It looks like we also need to move the $S_n$-reps class to a category so we don’t have to do a bunch of stupid subclasses, but I can do that on a separate PR.

jacksonwalters committed:
constructed vector space for InducedRepresentation

moved InducedRepresentation code from symmetric_group_algebra

jacksonwalters committed:
fix issues with RestrictedRepresentation

this is attempt to clean up the AI generated code for RestrictedRepresentation

jacksonwalters committed:
begin fixing TensorProductRepresentation

jacksonwalters committed:
begin fixing Frobenius reciprocity

jacksonwalters committed:
begin revising RestrictedRepresentation

jacksonwalters committed:
fix missing `_on_basis` definition

defines a default _on_basis function that acts on a basis element by embedding the subgroup element into the ambient group and delegating to the super-representation's action

jacksonwalters committed:
define dimension in InducedRepresentation

jacksonwalters committed:
slim down RestrictedRepresentation class

jacksonwalters committed:
extend tensor product to multiple factors

jacksonwalters committed:
update

jacksonwalters committed:
fix InducedRepresentation

jacksonwalters committed:
fix ExternalTensorProductRepresentation

jacksonwalters committed:
clean up restrict

jacksonwalters commented:
@tscrim Thanks for the detailed comments. I reworked the implementation along these lines:

- **External tensor product**
  - Replaced the previous specialized tensor product code with a more general `ExternalTensorProductRepresentation`.
  - It now supports an arbitrary number of factors.
  - The factors may be representations of different groups/semigroups.
  - The acting semigroup is the Cartesian product of the acting semigroups.
  - Added the convenience method `V.external_tensor_product(W, ...)`.

- **Restriction**
  - Reworked `RestrictedRepresentation` to be generic.
  - It now just changes the acting semigroup and delegates the action through the original representation.
  - It supports an optional `embedding` map.
  - The side of the original representation is preserved.
  - Added `V.restrict(H)` / `V.restriction(H)`.

- **Induction**
  - Reworked `InducedRepresentation` toward the general finite-group case.
  - It chooses coset representatives for `H` in `G`.
  - When possible, it uses `libgap.RightTransversal`; otherwise it falls back to enumerating the finite group.
  - It factors elements as `g = r h` with `r` a coset representative and `h in H`, caching those factorizations.
  - Added the convenience method `V.induce(G)`.

- **Coxeter/parabolic optimization**
  - Not implemented in this PR.
  - The current implementation is the generic finite-group version.
  - The minimal-length Coxeter/parabolic representative algorithm would be a good follow-up optimization.

- **The `S_k \subseteq S_n` / parabolic viewpoint**
  - Not implemented as a separate specialized interface yet.
  - The current code handles induction through the generic subgroup/coset mechanism.
  - Treating this explicitly as `S_k x S_1^{n-k}` should fit naturally into the later symmetric-group-specific layer.

- **Partition tuple input / `TabloidRepresentation`**
  - Not implemented in this PR.
  - I think this should be a follow-up on top of the generic induction/restriction/tensor product framework.
  - In particular, the partition tuple interface and `TabloidRepresentation` subclass still remain to be added.

- **Matrices**
  - Removed the need to implement matrices directly.
  - The new constructions define the semigroup action, and the existing base-class machinery handles representation matrices.

- **Dimension formula**
  - Implemented the induced representation dimension as
    `|G| / |H| * dim(V)`.
  - This avoids the incorrect binomial-index behavior for `S_k \subseteq S_n`.

- **Style / Sage conventions**
  - Cleaned up formatting and doctests.
  - Kept `libgap` imports local.
  - Made the Frobenius reciprocity checker private as `_verify_frobenius_reciprocity`.
  - Added `TestSuite` doctests for restriction, induction, and external tensor products.
  - Added tests for arbitrary-factor external tensor products and mixed left/right side rejection.

- **Separate category refactor**
  - I did not address moving the `S_n` representation class to a category in this PR.
  - That sounds like the separate follow-up you mentioned.
